wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -34,48 +37,58 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.annotation.PostConstruct;
 import javax.annotation.PreDestroy;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 import java.util.ArrayDeque;
-import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Queue;
 import java.util.concurrent.locks.ReadWriteLock;
-import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Predicate;
 
 import static java.lang.String.format;
 import static org.eclipse.che.dto.server.DtoFactory.newDto;
 
 /**
  * Defines an internal API for managing {@link WorkspaceRuntimeImpl} instances.
  *
- * <p>This component implements {@link WorkspaceStatus} spec.
+ * <p>This component implements {@link WorkspaceStatus} contract.
  *
  * <p>All the operations performed by this component are synchronous.
  *
- * <p>The implementation is thread-safe and guarded by {@link ReentrantReadWriteLock rwLock}.
+ * <p>The implementation is thread-safe and guarded by
+ * eagerly initialized readwrite locks produced by {@link WorkspaceRuntimes#STRIPED}.
+ * The component doesn't expose any api for client-side locking.
+ * All the instances produced by this component are copies of the real data.
  *
- * <p>The implementation doesn't validate parameters.
- * Parameters should be validated by caller of methods of this class.
+ * <p>The component doesn't check if the incoming objects are in application-valid state.
+ * Which means that it is expected that if {@link #start(WorkspaceImpl, String)} method is called
+ * then {@code WorkspaceImpl} argument is a application-valid object which contains
+ * all the required data for performing start.
  *
  * @author Yevhenii Voevodin
  * @author Alexander Garagatyi
  */
 @Singleton
 public class WorkspaceRuntimes {
 
-    private static final Logger LOG = LoggerFactory.getLogger(WorkspaceRuntimes.class);
+    private static final Logger                 LOG     = LoggerFactory.getLogger(WorkspaceRuntimes.class);
+    private static final Striped<ReadWriteLock> STRIPED = Striped.readWriteLock(16);
 
-    private final ReadWriteLock                         rwLock;
-    private final Map<String, RuntimeDescriptor>        descriptors;
-    private final Map<String, Queue<MachineConfigImpl>> startQueues;
-    private final MachineManager                        machineManager;
-    private final EventService                          eventService;
+    @VisibleForTesting
```

Comment:
This field is package published, which means that clients can use it, when we put this annotation on the field/method it is package visible but, it says this field should be used directly as it it exposed for tests. 

> Annotates a program element that exists, or is more widely visible than
> otherwise necessary, only for use in test code.